src/launch/service.c

```diff
@@ -398,20 +413,19 @@ static int service_watch_unit_handler(sd_bus_message *message, void *userdata, s
         return 0;
 }
 
-static int service_watch_unit(Service *service, const char *unit) {
-        _c_cleanup_(c_freep) char *object_path = NULL;
+static int service_watch_unit_callback(sd_bus_message *message, void *userdata, sd_bus_error *errorp) {
+        Service *service = userdata;
+        const char *object_path;
         int r;
 
-        assert(!service->slot_watch_unit);
+        if (sd_bus_message_get_error(message))
+                goto error;
 
-        r = sd_bus_path_encode(
-                "/org/freedesktop/systemd1/unit",
-                unit,
-                &object_path
-        );
-        if (r < 0)
-                return error_origin(r);
+        r = sd_bus_message_read(message, "o", &object_path);
+        if (error_trace(r) < 0)
+                goto error;
 
+        service->slot_watch_unit = sd_bus_slot_unref(service->slot_watch_unit);
```

Comment:
This should probably be at the top of the function to clean everything up and prepare for any followup. I don't think anything leaks if doing this late, but I think it would be easier to review if this is done early.